styx-scheduler-service/src/main/java/com/spotify/styx/state/QueuedStateManager.java

```diff
@@ -60,132 +59,203 @@
 
-  @Override
-  public void restore(RunState runState, long count) {
-    final WorkflowInstance workflowInstance = runState.workflowInstance();
-    if (states.containsKey(workflowInstance)) {
-      throw new RuntimeException("RunState initialization called on active instance "
-                                 + workflowInstance.toKey());
-    }
+    return CompletableFuture.supplyAsync(() -> {
 
-    states.computeIfAbsent(workflowInstance, (wfi) -> new InstanceState(wfi, runState, count + 1));
+      // Write active state to datastore
+      final RunState runState = RunState.create(workflowInstance, State.NEW, time.get());
+      try {
+        storage.runInTransaction(tx -> {
+          final Optional<Workflow> workflow = tx.workflow(workflowInstance.workflowId());
+          if (!workflow.isPresent()) {
+            throw new IllegalArgumentException("Workflow not found: " + workflowInstance.workflowId().toKey());
+          }
+          return tx.insertActiveState(workflowInstance, PersistentWorkflowInstanceState.of(runState, counter));
+        });
+      } catch (TransactionException e) {
+        if (e.isAlreadyExists()) {
+          throw new IllegalStateException("Workflow instance is already triggered: " + workflowInstance.toKey());
+        } else if (e.isConflict()) {
+          LOG.debug("Transactional conflict, abort triggering Workflow instance: " + workflowInstance.toKey());
+          throw new RuntimeException(e);
+        } else {
+          throw new RuntimeException(e);
+        }
+      } catch (IOException e) {
+        throw new RuntimeException(e);
+      }
+      return null;
+    }).thenCompose((nil) -> {
+      final Event event = Event.triggerExecution(workflowInstance, trigger);
+      try {
+        return receive(event);
+      } catch (IsClosedException isClosedException) {
+        LOG.warn("Failed to send 'triggerExecution' event", isClosedException);
+        // Best effort attempt to rollback the creation of the NEW state
+        try {
+          storage.deleteActiveState(workflowInstance);
+        } catch (IOException e) {
+          LOG.warn("Failed to remove dangling NEW state for: {}", workflowInstance);
+        }
+        throw new RuntimeException(isClosedException);
+      }
+    });
   }
 
   @Override
   public CompletionStage<Void> receive(Event event) throws IsClosedException {
     ensureRunning();
+    LOG.debug("Event {}", event);
 
-    final InstanceState state = states.get(event.workflowInstance());
-    if (state == null) {
-      String message = "Received event for unknown workflow instance: " + event;
-      LOG.warn(message);
-      return exceptionallyCompletedFuture(new IllegalArgumentException(message));
-    }
+    // TODO: optional retry on transaction conflict
 
-    CompletionStage<Void> processed = state.transition(event);
-    signalDispatcher();
-    return processed;
-  }
+    // TODO: run on striped executor to get event execution in order.
```

Comment:
How important is this? Sounds quite important, like a FIXME not just TODO? Or maybe we're fine as is, but I don't know what safeguard lets us claim that?